FUSETOOLS-1779: One change is applied more times in SAP components

apodhrad opened the pull request.

apupier commented on core/plugins/org.fusesource.ide.camel.model.service.core/src/org/fusesource/ide/camel/model/service/core/util/PropertiesUtils.java:
```diff
@@ -470,6 +479,29 @@ public static void updateURIParams(AbstractCamelModelElement selectedEP, Paramet
+     */
+    public static String replaceParts(String text, String target, String replacement, String delimiter) {
+    	String[] parts = text.split(delimiter);
+		for (int i = 0; i < parts.length; i++) {
```
indentation seems not correct

apodhrad replied:
it should be ok now (I didn't format the whole java code just to keep as small changes as possible)

apupier replied:
seems ok now :-)
it is effectively better to avoid changes everywhere only for formatting reasons.
To apply formatting only on few lines you can select the lines and then use Ctrl+Shift+F

apupier commented:
thank you very much for the PR!

We are trying to write tests as far as we can now. I think that it won't be too hard to add unit test in this case.
Can you add them please? I can help you if you need.
(if not possible Lars or I can also work on it)

apodhrad commented:
Unfortunately, I don't know how to easily create an instance of AbstractCamelModelElement :(

apupier commented:
Effectively this a major flaw for writing unit tests currently.
I think that you should extract 
String newWithoutScheme = replaceParts(withoutScheme, pparam.getName(), val, ":");
// if nothing happens then use the old logic
if (withoutScheme.equals(newWithoutScheme)) {
	withoutScheme = withoutScheme.replace(pparam.getName(), val);
} else {
	withoutScheme = newWithoutScheme;
}

in a dedicated method ( removeScheme() for instance) 
Like that you will get rid of the CamelModelModelElement and be able to write Unit tests for the old way and new way and ensure that both cases are working.

We can also write an integration tests, this time you can have a look to https://github.com/fusesource/fuseide/blob/25e8e18e75d2b67eec5191c17587148028a72bd8/core/tests/org.fusesource.ide.camel.validation.tests.integration/src/org/fusesource/ide/camel/validation/xml/XMLCamelRoutesValidatorIT.java

having both is better but if there are only unit tests for this case it would be enough for me

apodhrad commented:
I added a very basic test for the logic I changed. I will try to add an integration test. Once I'm done I will let you know. So, please do not push this PR at the moment.

apupier commented on core/tests/org.fusesource.ide.camel.model.service.core.tests/src-test/org/fusesource/ide/camel/model/service/core/util/PropertiesUtilsTest.java:
```diff
@@ -0,0 +1,83 @@
+package org.fusesource.ide.camel.model.service.core.util;
```
don't forget to provide the header also for test files

apupier commented on core/tests/org.fusesource.ide.camel.model.service.core.tests/src-test/org/fusesource/ide/camel/model/service/core/util/PropertiesUtilsTest.java:
```diff
@@ -0,0 +1,83 @@
+		expected = "sap-idoclist-destination:myDest:idocType:idocTypeExtension:systemRelease:applicationRelease";
+		assertEquals(expected, updatePathParams(syntax, pathParam("destination"), "myDest", pathParams, modelMap));
+
+		expected = "sap-idoclist-destination:destination:abc:idocTypeExtension:systemRelease:applicationRelease";
```
you should better write separated unit tests for each conditions.
It helps to have a concise view of what is failing exactly
You can @Before method for the initialization or just extract a method
It also helps to understand each tested case as we need to provide a new test method name for each case :-)

apupier commented:
Thank you for adding the unit tests, you extracted the method at good level i think (better than what I proposed initially ^^)
No problem to wait for the integration test

apodhrad commented:
during writing the integration test I found another bug in the logic :(

apupier commented:
That is the nice side of writing tests :-)

apodhrad commented:
Added also the integration test and fixed the other issue (the problem occured also if there was a slash in the uri)

apupier commented on core/tests/org.fusesource.ide.camel.model.service.core.tests/src-test/org/fusesource/ide/camel/model/service/core/util/PropertiesUtilsTestIT.java:
```diff
@@ -0,0 +1,116 @@
+package org.fusesource.ide.camel.model.service.core.util;
```
don't forget the header on the test file here too

apodhrad commented:
done

apupier commented on core/tests/org.fusesource.ide.camel.model.service.core.tests/src-test/org/fusesource/ide/camel/model/service/core/util/PropertiesUtilsTest.java:
```diff
@@ -0,0 +1,123 @@
+/*******************************************************************************
+ * Copyright (c) 2014 Red Hat, Inc.
```
we use the creation year, so 2016 instead of 2014

apodhrad commented:
changed to 2016

apodhrad commented:
I added one small improvements + one test for using regex. I think that the PR is done. If some changes are needed pls let me know.

apupier commented on core/tests/org.fusesource.ide.camel.model.service.core.tests/src-test/org/fusesource/ide/camel/model/service/core/util/PropertiesUtilsTestIT.java:
```diff
@@ -0,0 +1,126 @@
+ *
+ */
+@RunWith(Parameterized.class)
+public class PropertiesUtilsTestIT {
```
The It test need to be placed in another plugin: org.fusesource.ide.camel.model.service.core.tests.integration

apupier commented:
Thank you for your PR!
